trieste/models/gpflow/models.py

```diff
@@ -1362,3 +1368,259 @@ def covariance_between_points(
+        """
+        (
+            query_points_wo_fidelity,
+            query_points_fidelity_col,
+        ) = check_and_extract_fidelity_query_points(
+            query_points, max_fidelity=self.num_fidelities - 1
+        )
+
+        signal_mean, signal_var = self.lowest_fidelity_signal_model.predict(
+            query_points_wo_fidelity
+        )
+
+        for fidelity in range(self.num_fidelities):
+            if fidelity == 0:
+                fidelity_residual_mean = 0
+                fidelity_residual_var = 0
+            else:
+                fidelity_residual_mean, fidelity_residual_var = self.fidelity_residual_models[
+                    fidelity
+                ].predict(query_points_wo_fidelity)
+
+            new_fidelity_signal_mean = self.rho[fidelity] * signal_mean + fidelity_residual_mean
+            new_fidelity_signal_var = fidelity_residual_var + (self.rho[fidelity] ** 2) * signal_var
+
+            mask = query_points_fidelity_col >= fidelity
+            signal_mean = tf.where(mask, new_fidelity_signal_mean, signal_mean)
+            signal_var = tf.where(mask, new_fidelity_signal_var, signal_var)
+
+        return signal_mean, signal_var
+
+    def _calculate_residual(self, dataset: Dataset, fidelity: int) -> TensorType:
+        """
+        Calculate the true residuals for a set of datapoints at a given fidelity.
+
+        :param dataset: Dataset of points for which to calculate the residuals.
+            Query points shape is [N, D], observations is [N,P].
+        :param fidelity: The fidelity for which to calculate the residuals
+        :return: The true residuals at given datapoints for given fidelity, shape is [N,1].
+        """
+        fidelity_query_points = convert_query_points_for_fidelity(
+            dataset.query_points, fidelity - 1
+        )
+        residuals = (
+            dataset.observations - self.rho[fidelity] * self.predict(fidelity_query_points)[0]
+        )
+        return residuals
+
+    def sample(self, query_points: TensorType, num_samples: int) -> TensorType:
+        """
+        Sample `num_samples` samples from the posterior distribution at `query_points`
+
+        :param query_points: The query points at which to sample of shape [N, D+1], where the
+            final column of the final dimension contains the fidelity of the query point
+        :param num_samples: The number of samples (S) to generate for each query point.
+        :return: samples from the posterior of shape [S, N, P]
+        """
+        (
+            query_points_wo_fidelity,
+            query_points_fidelity_col,
+        ) = check_and_extract_fidelity_query_points(
+            query_points, max_fidelity=self.num_fidelities - 1
+        )
+
+        signal_sample = self.lowest_fidelity_signal_model.sample(
+            query_points_wo_fidelity, num_samples
+        )
+
+        for fidelity in range(self.num_fidelities):
```

Comment:
same as predict, going through all fidelities and query points is going to be very expensive.